src/modulus_convert.rs

```diff
@@ -0,0 +1,548 @@
+
+        assert_eq!(sub_mod_p(10u64, 1u64, 18_446_744_073_709_551_557_u64), 9u64,);
+
+        assert_eq!(
+            sub_mod_p(0u64, 1u64, 18_446_744_073_709_551_557_u64),
+            18_446_744_073_709_551_556_u64,
+        );
+
+        assert_eq!(
+            sub_mod_p(0u64, 59u64, 18_446_744_073_709_551_557_u64),
+            18_446_744_073_709_551_498_u64,
+        );
+
+        assert_eq!(
+            sub_mod_p(0u64, 60u64, 18_446_744_073_709_551_557_u64),
+            18_446_744_073_709_551_497_u64,
+        );
+
+        assert_eq!(
+            sub_mod_p(0u64, 61u64, 18_446_744_073_709_551_557_u64),
+            18_446_744_073_709_551_496_u64,
+        );
+    }
+
+    #[test]
+    fn test_mult_mod_p() {
+        assert_eq!(mult_mod_p(0u64, 0u64, 31u64), 0u64);
+        assert_eq!(mult_mod_p(1u64, 1u64, 31u64), 1u64);
+        assert_eq!(mult_mod_p(3u64, 7u64, 31u64), 21u64);
+        assert_eq!(mult_mod_p(5u64, 7u64, 31u64), 4u64);
+        assert_eq!(mult_mod_p(30u64, 7u64, 31u64), 24u64);
+        assert_eq!(mult_mod_p(30u64, 30u64, 31u64), 1u64);
+
+        assert_eq!(mult_mod_p(0u64, 0u64, 18_446_744_073_709_551_557_u64), 0u64,);
+        assert_eq!(mult_mod_p(0u64, 1u64, 18_446_744_073_709_551_557_u64), 0u64,);
+        assert_eq!(mult_mod_p(1u64, 1u64, 18_446_744_073_709_551_557_u64), 1u64,);
+        assert_eq!(
+            mult_mod_p(53u64, 9u64, 18_446_744_073_709_551_557_u64),
+            477u64,
+        );
+
+        assert_eq!(
+            mult_mod_p(
+                4_294_967_295_u64,
+                4_294_967_296_u64,
+                18_446_744_073_709_551_557_u64
+            ),
+            18_446_744_069_414_584_320_u64,
+        );
+        assert_eq!(
+            mult_mod_p(
+                4_294_967_296_u64,
+                4_294_967_296_u64,
+                18_446_744_073_709_551_557_u64
+            ),
+            59u64,
+        );
+        assert_eq!(
+            mult_mod_p(
+                2u64,
+                9_223_372_036_854_775_804_u64,
+                18_446_744_073_709_551_557_u64
+            ),
+            51u64,
+        );
+        assert_eq!(
+            mult_mod_p(
+                10_000_000_000_000_000_000_u64,
+                10_000_000_000_000_000_000_u64,
+                18_446_744_073_709_551_557_u64
+            ),
+            6_932_391_181_562_104_841_u64,
+        );
+        assert_eq!(
+            mult_mod_p(
+                14_140_769_320_479_119_572_u64,
+                16_523_972_853_534_698_712_u64,
+                18_446_744_073_709_551_557_u64
+            ),
+            1_263_076_288_923_396_945_u64,
+        );
+    }
+
+    #[test]
+    fn test_three_helpers() {
+        // let p = 2147483647; // 2^31 - 1
+        // let p = 2305843009213693951; // 2^61 - 1
+        let p = 18_446_744_073_709_551_557; // 2^64 - 59
+        let batch_size = 100;
+
+        // Helper 1 and 3 will share this seed
+        let seed_1: [u8; 32] = [1; 32];
```

Comment:
might consider slightly more compact
```suggestion
        let seed_1 = [1u8; 32];
```